File: android/app/src/main/kotlin/com/goalmaker/app/data/sync/SupabaseChangeFeed.kt
package com.goalmaker.app.data.sync

import com.goalmaker.app.domain.sync.SyncedTableCatalog
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.realtime.PostgresAction
import io.github.jan.supabase.realtime.RealtimeChannel
import io.github.jan.supabase.realtime.channel
import io.github.jan.supabase.realtime.postgresChangeFlow
import io.github.jan.supabase.realtime.realtime
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.awaitCancellation
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Listens to Supabase Realtime for changes to the synced tables while the app is visible, and
 * reports each as a nudge; the payload is never applied, the next pull fetches the rows
 * (docs/sync.md). Every (re)join is a nudge too, since changes made while disconnected produced no
 * events. Failures are quiet: sync still runs on writes, on start and in the background.
 */
class SupabaseChangeFeed(
    private val client: SupabaseClient,
    private val catalog: SyncedTableCatalog,
    private val scope: CoroutineScope,
    private val onChange: () -> Unit,
) {
    private val lock = Any()
    private var listening: Job? = null

    fun start() {
        synchronized(lock) {
            if (listening?.isActive != true) {
                listening = scope.launch { listen() }
            }
        }
    }

    fun stop() {
        synchronized(lock) {
            listening?.cancel()
            listening = null
        }
    }

    private suspend fun listen() {
        val channel = client.channel(CHANNEL)
        try {
            coroutineScope {
                for (synced in catalog.tables) {
                    channel.postgresChangeFlow<PostgresAction>(schema = "public") { table = synced.name }
                        .onEach { onChange() }
                        .launchIn(this)
                }
                channel.status
                    .filter { it == RealtimeChannel.Status.SUBSCRIBED }
                    .onEach { onChange() }
                    .launchIn(this)
                channel.subscribe(blockUntilSubscribed = false)
                awaitCancellation()
            }
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            // Realtime is a speed-up only.
        } finally {
            withContext(NonCancellable) {
                runCatching { client.realtime.removeChannel(channel) }
            }
        }
    }

    private companion object {
        const val CHANNEL = "goalmaker-sync"
    }
}

File: windows/src/GoalMaker.Infrastructure/Sync/SupabaseChangeFeed.cs
using GoalMaker.Core.Sync;
using Supabase.Realtime;
using Supabase.Realtime.PostgresChanges;

namespace GoalMaker.Infrastructure.Sync;

/// <summary>
/// Listens to Supabase Realtime for changes to the synced tables and reports each as a nudge; the
/// payload is never applied, the next pull fetches the rows (docs/sync.md). Every (re)join is a nudge
/// too, since changes made while disconnected produced no events. Failures are quiet: the periodic
/// sync still runs.
/// </summary>
public sealed class SupabaseChangeFeed(Supabase.Client client, SyncedTableCatalog catalog, Action onChange) : IAsyncDisposable
{
    private RealtimeChannel? channel;

    public async Task StartAsync(string accessToken)
    {
        await StopAsync().ConfigureAwait(false);
        try
        {
            client.Realtime.SetAuth(accessToken);
            await client.Realtime.ConnectAsync().ConfigureAwait(false);
            var joined = client.Realtime.Channel("goalmaker-sync");
            foreach (var table in catalog.Tables)
            {
                joined.OnPostgresChange(
                    (_, _) => onChange(),
                    PostgresChangesOptions.ListenType.All,
                    new PostgresChangesFilter { Schema = "public", Table = table.Name });
            }

            joined.AddStateChangedHandler((_, state) =>
            {
                if (state == Constants.ChannelState.Joined)
                {
                    onChange();
                }
            });
            await joined.Subscribe().ConfigureAwait(false);
            channel = joined;
        }
        catch (Exception error) when (error is not OutOfMemoryException)
        {
            // Realtime is a speed-up only; sync still runs on writes and on its timer.
        }
    }

    public void UpdateToken(string accessToken)
    {
        try
        {
            client.Realtime.SetAuth(accessToken);
        }
        catch (Exception error) when (error is not OutOfMemoryException)
        {
            // Same as above.
        }
    }

    public async Task StopAsync()
    {
        if (channel is null)
        {
            return;
        }

        try
        {
            channel.Unsubscribe();
            client.Realtime.Remove(channel);
            client.Realtime.Disconnect();
        }
        catch (Exception error) when (error is not OutOfMemoryException)
        {
            // Already gone.
        }

        channel = null;
        await Task.CompletedTask.ConfigureAwait(false);
    }

    public ValueTask DisposeAsync() => new(StopAsync());
}
